### PR TITLE
feat(stdlibs): sha3.Sum256(), sha3.Sum512()

### PR DESCRIPTION
**Pull request by r3v4s**
# Description

As bit of some discussions on #570, #580 
Strong hash algorithm need in gno.

AFAIK, currently NIST's standard is SHA3 which is family of FIPS 180-4 and FIPS 202 ([ref](https://www.nist.gov/publications/secure-hash-standard)).

# How has this been tested?

tests by gnodev

### CODE REVIEW

**Comment by r3v4s** [comment]
@moul I've resolve conflict issue with latest master branch.
Can it be merge just like #580 ??

File: stdlibs/stdlibs.go [hunk 1]
```diff
@@ -21,6 +24,59 @@ func InjectNativeMappings(store gno.Store) {
...
+			func(m *gno.Machine) {
+				arg0 := m.LastBlock().GetParams1().TV
+				bz := []byte(nil)
+				if arg0.V != nil {
+					slice := arg0.V.(*gno.SliceValue)
+					array := slice.GetBase(m.Store)
+					bz = array.GetReadonlyBytes()[:slice.Length]
+				}
+				// remove padding
+				bzNP := bytes.Trim(bz, "\x00")
```
Comment (thehowl) [inline stdlibs/stdlibs.go:44]: Why are bytes padded with zero bytes in the first place here? It seems wrong to do this... (we should hash alll bytes in the input)

**Review by thehowl — CHANGES_REQUESTED** [review]

File: stdlibs/stdlibs.go (hunk 1 above)
Comment (r3v4s) [reply stdlibs/stdlibs.go:44]: that was an old legacy when I didn't slice the result of GetReadonlyBytes().
just forgot to remove it :D

**Review by thehowl — COMMENTED** [review]
Aside from needing to merge and resolve conflicts with master, I think this PR requires a review/approval from @moul (per discussion on #679), as well as a bit more discussion on the feature per se...

I think there is good reason why sha3 was not included in the standard library, I've rarely ever seen it used out in the wild (I don't know if it has specific uses in crypto-land, if so please enlighten me). I would rather not add packages to the standard library if they don't have an obvious use case, especially if the hashing function is comparable in performance and quality to sha256.

If necessary, I think from the [source code of x/crypto/sha3](https://cs.opensource.google/go/x/crypto/+/master:sha3/) that it could be implemented in pure gno (although we may need to add `RotateLeft64` to `math/bits`, which keccakf.go makes extensive use of). And to me that would be somehow better if somebody needs to use it?

**Comment by moul** [comment]
The initial version of GnoVM (mainnet) will be lightweight. We are currently considering the inclusion of native types as a proof of concept and to consider improving the preprocessor (see #650). However, for existing libraries in Go, like the one proposed in this PR, we will ask for a pure userland implementation first. In the future, GnoVM may specialize in certain areas based on a DAO vote, taking into account the benefits. Regarding this specific package, the goal is to have numerous use cases that could justify making it native for potential performance improvements, although I highly doubt it will ever occur.

Let's establish a practice for userland libraries, such as this one, to submit PRs to this repository, adding them in `examples/gno.land/p/...`. Additionally, we should find a way to manage licensing effectively for ported content from non-Gno authors on GitHub. It is crucial to make library porting a common and straightforward process. See https://github.com/gnolang/gno/issues/239.

Maybe we should make a `gno.land/p/port` folder so we can tne make `gno.land/p/port/github.com/gnolang/x/crypto`?

**Comment by r3v4s** [comment]
@moul Could be more specific about what do you mean by `pure userland implementation`?
> Do you mean we should implement those packages with `gnolang` without `go native binding`?

**Comment by thehowl** [comment]
@r3v4s I'll close this PR to do some housekeeping on our old PRs.

Feel free to create a new PR to add package `examples/gno.land/p/demo/crypto/sha3`. Maybe consider adding also `ShakeSum*` functions, as those are the ones recommended by the package documentation itself:

>The SHAKE functions are recommended for most new uses. They can produce output of arbitrary length. SHAKE256, with an output length of at least 64 bytes, provides 256-bit security against all attacks. The Keccak team recommends it for most applications upgrading from SHA2-512

**Pull request closed without merging**